cryptoprotocol/src/main/java/org/cryptimeleon/incentivesystem/cryptoprotocol/proof/SpendDeductZkp.java

```diff
@@ -0,0 +1,164 @@
+        // c1=esk*gamma+dsrnd1
+        var c1statement = new ExponentSumExpr(new ExponentMulExpr(eskVar, new ExponentConstantExpr(commonInput.gamma)), dsrnd1Var);
+        builder.addSubprotocol("c1=esk*gamma+dsrnd1", new LinearExponentStatementFragment(c1statement.isEqualTo(commonInput.c1), zn));
+
+        // dsid=w^esk
+        var dsidEskStatement = w.pow(eskVar).isEqualTo(commonInput.dsid);
+        builder.addSubprotocol("dsid=w^esk", new LinearStatementFragment(dsidEskStatement));
+
+        // C=(H.pow(usk, esk, dsrnd_0, dsrnd_1, v, z, t),g_1) split into two subprotocols
+        var commitmentC0Statement = H.pow(new Vector<>(
+                uskVar,
+                eskVar,
+                dsrnd0Var,
+                dsrnd1Var,
+                vVar,
+                zVar,
+                tVar
+        )).reduce(GroupElementExpression::op).isEqualTo(commonInput.commitmentC0);
+        builder.addSubprotocol("C0", new LinearStatementFragment(commitmentC0Statement));
+        builder.addSubprotocol("C1", new LinearStatementFragment(commonInput.commitmentC1.expr().isEqualTo(pp.getG1())));
+
+        // C=(H.pow(usk, \sum_i=0^k[esk^*_(usr,i) * base^i], dsrnd^*_0, dsrnd^*_1, v-k, z^*, t^*), g_1^(u^*)) split into two subprotocols
+        // We use the sum to combine the esk^*_usr = \sum proof with the C=.. proof
+        var exponents = new Vector<>(uskVar, BigInteger.ZERO, dsrndStar0Var, dsrndStar1Var, vVar.sub(zn.valueOf(commonInput.k)), zStarVar, tStarVar); // 0 because h_2^(esk^*_usr) is handled separately
+        GroupElementExpression eskPart = groupG1.getNeutralElement().expr();
+        for (int i = 0; i < pp.getNumEskDigits(); i++) {
+            eskPart = eskPart.op(H.get(1).pow(eskDecVarVector[i].mul(pp.getEskDecBase().pow(BigInteger.valueOf(i)))));
+        }
+
+        var cPre0Statement = H.pow(exponents).reduce(GroupElementExpression::op).op(eskPart).isEqualTo(commonInput.c0Pre.pow(uStarInverseVar));
+        builder.addSubprotocol("C0Pre", new LinearStatementFragment(cPre0Statement));
+        builder.addSubprotocol("C1Pre", new LinearStatementFragment(commonInput.c1Pre.expr().isEqualTo(pp.getG1().pow(uStarVar))));
+        builder.addSubprotocol("u^*-inverse", new LinearStatementFragment(commonInput.c1Pre.pow(uStarInverseVar).isEqualTo(pp.getG1()))); // Prove that the inverse is actually the inverse
+
+        // esk^*_(usr,i)\in[base]
+        for (int i = 0; i < pp.getNumEskDigits(); i++) {
+            builder.addSubprotocol("eskDigitSetMembership_" + i, new SetMembershipFragment(pp.getEskBaseSetMembershipPublicParameters(), eskDecVarVector[i]));
+        }
+
+        // v >= k (I have more points than required)
+        // We prove that v-k\in[0,eskDecBase^{maxPointBasePower+1}-1]
+        builder.addSubprotocol("v>=k", new SmallerThanPowerFragment(vVar.sub(new ExponentConstantExpr(commonInput.k)), pp.getEskDecBase().getInteger().intValue(), pp.getMaxPointBasePower(), pp.getEskBaseSetMembershipPublicParameters()));
```

Comment:
Replace the `new ExponentConstantExpr` with `.asExponentExpr`.

Reply:
Currently, k is a `BigInteger`, I could either change that to ZnElement, or leave it the way it is

Reply:
Ah. Nevermind. Anyway, https://github.com/cryptimeleon/math/pull/99 should make the `new ExponentConstantExpr` redundant.